Misc get started tweaks and more robust `useLazyQuery` documentation (#12833)
apollographql/apollo-client: docs/source/data/queries.mdx

Author: Note: I've avoided using `dataState` in the examples until we have that completely flushed out.

</Note>

#### Changing variables

You change variables by calling the execution function with updated variables.

The following is an example that gets the selected dog's photo when clicking the "Get photo" button.

```jsx
function DogPhoto() {
  const [selectedDog, setSelectedDog] = useState(null);
  const [getDogPhoto, { loading, error, data }] = useLazyQuery(GET_DOG_PHOTO);

  if (loading) return <p>Loading ...</p>;
  if (error) return `Error! ${error.message}`;

  return (
    <div>
      <DogSelect
        onChange={(dogName) => setSelectedDog(dogName)}
        value={selectedDog}
      />
      {data?.dog && <img src={data.dog.displayImage} />}
      {selectedDog && (
        <button
          onClick={() => getDogPhoto({ variables: { name: selectedDog } })}
        >
          Get photo
        </button>
      )}
    </div>
  );
}
```

If you need to reference the currently executed query's variables, use the `variables` property returned by `useLazyQuery`. The `variables` property contains the value of the variables from the last execution of the query.

<Tip>

> The `variables` property is empty until the execution function is called for the first time. Use the `called` property returned by `useLazyQuery` to determine when the execution function has been called at least once.

Reviewer: ```suggestion
The `variables` property is empty until the execution function is called for the first time. Use the `called` property returned by `useLazyQuery` to determine if the execution function has been called at least once.
```

  return (
    <div>
      {/* ... */}
      <button onClick={() => getDogPhoto({ variables: { name: selectedDog } })}>
        Get photo
      </button>

      {called && <span>Last fetched: {variables.name}</span>}
    </div>
  );
}
```

### Using the promise returned from the execution function

The execution function returns a promise that resolves with the query result:

```jsx
function GetDogs() {
  const [getDogs, { data, loading, error }] = useLazyQuery(GET_DOGS);

  const handleClick = async () => {
    const { data } = await getDogs();

    // Do something with `data`
  };

  return (
    <div>
      {/* ... */}
      <button onClick={handleClick}>Get dogs</button>
    </div>
  );
}
```

The result returned from the promise is useful when you need to execute side-effects using the data returned by the query.

<Tip>

Use the `data`, `error`, and other properties returned by `useLazyQuery` to sync the query state with your component. Avoid using your own state setters from React's `useState` hook with data resolved from the promise. This ensures your component stays up-to-date with cache changes as they occur throughout your application.<br /><br />

> In cases where you don't need to keep your component in sync with query state, use `client.query()` directly because it won't unnecessarily render your component for data that you don't use.
> 
> <ExpansionPanel title="Example">

Reviewer: ```suggestion
In cases where you don't need to keep your component in sync with query state, use `client.query()` directly because it won't unnecessarily render your component for data that you don't use.

Using `useLazyQuery` in those situations should be considered an antipattern.

<ExpansionPanel title="Example">
```

Let's really try to drive this point home.

```ts
const handleClick = async () => {
  const { data, error } = await getDogs();

  if (error && CombinedGraphQLErrors.is(error)) {
    // handle GraphQL errors returned by the query
  }
};
```

<Note>

`data` might not contain any partial data and is instead set to `undefined`. This typically occurs when a [network error](./error-handling#network-errors) causes the query to fail because the error might not be associated with GraphQL execution.

</Note>

##### `errorPolicy: "ignore"`

The promise resolves with an object that includes any partial data returned by the query. Errors are discarded.

```ts
const handleClick = async () => {
  const { data } = await getDogs();

  if (data !== undefined) {
    // Do something with the returned data
  }
};
```

<Tip>

`data` might be `undefined` in the event a [network error](./error-handling#network-errors) is raised. We recommend checking if `data` is `undefined` before attempting to use it in case an error occurs during query execution.

</Tip>

#### Retaining query results

> In-flight queries executed by `useLazyQuery` are aborted when the component unmounts, causing the promise to reject. In some cases, you might find this behavior undesirable and would prefer to let the query run to completion.

Reviewer: ```suggestion
In-flight queries executed by `useLazyQuery` are aborted when the component unmounts or another query ist started via the execution function, causing the promise to reject. (This is a silent rejection, so unless you called `.then`, `.catch` or `await`ed the promise, you won't see it.)
```

Reviewer: Two edits in this - the addition in parantheses could also be a `<Note` pararaph? Not sure.

> The promise returned by the execution function includes a `.retain()` method. When called, it ensures the query continues running even when the component unmounts.

Reviewer: ```suggestion
The promise returned by the execution function includes a `.retain()` method. When called, it ensures the query continues running even when the component unmounts or a new query is started before the last one finished.
```

Reviewer: I noticed we didn't have a test for that, so I added one in cbe685afb87fb9fcacb22429521e4972bd9634cf

Author: Good catch! Thanks

#### Retaining query results

In-flight queries executed by `useLazyQuery` are aborted when the component unmounts or another query is started when calling the execution function, causing the promise to reject. In some cases, you might find this behavior undesirable and would prefer to let the query run to completion.

<Note>

> Apollo Client ensures the rejected promise doesn't throw an unhandled rejection error when you don't add a rejection handler to the promise. This however means that aborted errors are silent and might go unnoticed. If you want to be notified when the request is aborted, provide a rejection handler for the promise.

Reviewer: This is not completely correct - calling `promise.then(onlyAResolvedHandler)` will also cause an error to be thrown, not only `promise.then(resolvedHandler, rejectionHandler)` or `promise.catch(rejectionHandler)`
